### PR TITLE
[Placement Group] [Test] Fix pg.ready hang forever when gcs restarting

### LINKED ISSUE #20058: [Bug] [Placement Group] [Test]  pg.ready() hang forever when GCS server restarting.
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

`test_placement_group_3.py#test_create_placement_group_during_gcs_server_restart` flaky on master branch.
currently, it was skipped.

### Versions / Dependencies

None.

### Reproduction script

None.

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by clay4megtr**
## Why are these changes needed?

fix the failed ut test_placement_group_3.py::test_create_placement_group_during_gcs_server_restart.

the flaky cases were like this:
1. register a placement group and start to schedule.
2. the creation will fail because of the GCS restarting.
3. GCS restart and send cancel resource request to raylets to release the reserved resource before. note here, **when raylet remove the reserved bundle, it will send an RPC to GCS to notify GCS that there is a resource removed event though the `delete_resources_` function in `placement_group_resource_manager.h`**
4. reschedule the placement group,  send prepare and commit RPC requests to raylet, then, the raylet end recreate the placement group resources and notify GCS that there is a resource add event though the `update_resources_` function in `placement_group_resource_manager.h`
5.  **Raylet receives the `resource removed` event now, so it will remove the placement group resource that we just successfully created**.
6. Then, Raylet receive the `resource added` event right now, this time, **it only adds the placement group resource to its local resource view including all nodes but does not add pg resources to `local_resources_` defined in `cluster_resource_scheduler.cc`**
7. The `ready` task submitted and queued, it will hang forever as there aren't the placement group resource  in `local_resources_`.

## Related issue number

Closes #20058 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
> it only adds the placement group resource to its local resource view including all nodes but does not add pg resources to local_resources_ defined in cluster_resource_scheduler.cc

Can you actually explain this part a little bit? What's the difference between `local_resources_` and `local resource view`?

**Comment by rkooo567** [comment]
Also, is this something that can happen without gcs restarting too? Your description sounds like that, but I'd like to get confirmation

**Comment by rkooo567** [comment]
One last thing I am confused is; 

```
Raylet receives the resource removed event now, so it will remove the placement group resource that we just successfully created.
Then, Raylet receive the resource added event right now, this time, it only adds the placement group resource to its local resource view including all nodes but does not add pg resources to local_resources_ defined in cluster_resource_scheduler.cc
```

Why do we update the local resource view based on these events? Aren't raylet supposed to already update local resources when the resources are added / deleted?

**Comment by clay4444** [comment]
okay, let me add more details, when raylet starts, it starts to register a node resource changed listener to GCS to listen the resource changed event of all nodes, so that it can sense the resource change of all nodes and update the resource view of `cluster_resource_scheduler.cc`. 

`cluster_resource_scheduler` consists of two sorts of resource view, one is `nodes_` map container whose key is node id and value is `NodeResources` (**actually the `local resource view` I mentioned before is this map container**). the other is `local_resources_` whose type is `NodeResourceInstances`. 

when raylet receives the resource deleted event, it invokes the `cluster_resource_scheduler_->DeleteResource` to delete resources from `cluster_resource_scheduler.cc`, it actually deletes resources from both `nodes_` and `local_resources_`.  however, when raylet receives the resource update event, it invokes the `cluster_resource_scheduler_->UpdateResourceCapacity` to update resources from `cluster_resource_scheduler.cc`,  **this time it only updates resources from `nodes_` map container but does not update the `local_resources_`.** this is actually the root cause!

**Comment by clay4444** [comment]
then, let's see why this trigger the bug:

1. GCS restart. it re-schedules the placement group that has prepared the resource but has not committed resource before and sends RPC requests to raylets to cancel the prepared resources before.
2. Raylet receives the cancel request and cancels the prepared resources, after this, it invokes `delete_resources_ ` to send a delete resource RPC request to GCS.
3. GCS `GcsResourceManager::HandleDeleteResources` be invoked to handle the request, **Also it publishes a resource delete event**
4. Continue to re-schedule the placement group, GCS end sends prepare and commit RPC requests to raylet to re-create the placement group.
5. Raylet **re-create** the placement group, add pg resources to `cluster_resource_scheduler.cc` and invoke `update_resources_` to send an update resource RPC request to GCS.
6. GCS `GcsResourceManager::HandleUpdateResources` be invoked to handle the request, **Also it publishes a resource update event**
7. Raylet receives the `resource deleted` event, it invokes `cluster_resource_scheduler_->DeleteResource` method to delete resources from `cluster_resource_scheduler.cc`, **This will delete the pg resource that we just created at step 5**.  **Note that this will delete resources both from `nodes_ ` and `local_resources_ `.**
8. Raylet receives the `resource update` event, it invokes `cluster_resource_scheduler_->UpdateResourceCapacity`  to update resources from `cluster_resource_scheduler.cc`, **however, this time it will only add pg resources to `nodes_`**
9. Up to this point, you can find that we have re-created the placement group successfully. but the `local_resources_ ` created in the `cluster_resource_scheduler.cc` does not contain the pg resources.

**Comment by clay4444** [comment]
Also, let's see the scheduler of the `pg.ready()` task.

1. We use `cluster_resource_scheduler_->GetBestSchedulableNode` to judge whether the task can be scheduled or not(you can check this in `ClusterTaskManager::SchedulePendingTasks`), in this case, the result is True and we can successfully get a node to schedule this task because there indeed has the placement group resource in the `nodes_`.
2. However, when we try to dispatch this task to a worker, we use `cluster_resource_scheduler_->AllocateLocalTaskResources` to allocate resources for this task (you can check this in `ClusterTaskManager::DispatchScheduledTasksToWorkers`), this will fail because it will allocate resources from `local_resources_ ` and the `local_resources_ ` does not contain the pg resources right now as I mentioned before.

**Comment by rkooo567** [comment]
So, it seems like the issue is that deletion event is published after we commit resources, which deletes resources from the local resources. And your fix is to update local resources again when new creation resources event is published, am I correct?

In this case, why don't we just ignore the deprecated delete event if the deleted bundle is from the local node? Theoretically, local raylet should have the "most accurate" resource information, and updating local resources based on GCS event seems to be vulnerable to race condition like this. Your approach will fix one of potential problems, but as long as we rely on local resources information from GCS, there could be a lot more unexpected race condition.

**Comment by clay4444** [comment]
Yeah, exactly correct!   and that's a good idea!  I'll update this.

**Comment by clay4444** [comment]
please take a look again! @rkooo567

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 1]
```diff
@@ -168,6 +168,13 @@ class ClusterResourceScheduler : public ClusterResourceSchedulerInterface {
                               const std::string &resource_name,
                               double resource_total) override;
 
+  /// Re-init the capacity of the local resources.
+  ///
+  /// \param resource_name: Resource which we want to update.
+  /// \param resource_total: New capacity of the resource.
+  void UpdateLocalResourceCapacity(const std::string &resource_name,
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:175]: Do we still need this in this case? If the delete events don’t remove local resources, we don’t need to re-init, or am I missing sth here?

Comment (clay4444) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.h:175]: yeah, you are right, we can remove the previous code safely now.

**Comment by rkooo567** [comment]
Nice! Btw, is this testable? (in cpp)

Also, should we try running tests like 5~6 times before we merge? (to make sure it is not flaky). You can achieve in this way; https://github.com/ray-project/ray/blob/dc51af798c3cdb79aceaf9f0c1ca63fc1d2b12b5/python/ray/tests/test_placement_group_mini_integration.py#L123

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Windows failure seems to be a false signal

**Pull request merged**